pennylane/templates/subroutines/all_singles_doubles.py

```diff
@@ -0,0 +1,204 @@
+        doubles (Sequence[Sequence]): sequence of lists with the indices of the four qubits
+            the :class:`~.pennylane.DoubleExcitation` operations act on
+
+    .. UsageDetails::
+
+        Notice that:
+
+        #. The number of wires has to be equal to the number of spin orbitals included in
+           the active space.
+
+        #. The single and double excitations can be generated with the function
+           :func:`~.excitations`. See example below.
+
+        An example of how to use this template is shown below:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            import numpy as np
+
+            electrons = 2
+            qubits = 4
+
+            # Define the HF state
+            hf_state = qml.qchem.hf_state(electrons, qubits)
+
+            # Generate all single and double excitations
+            singles, doubles = qml.qchem.excitations(electrons, qubits)
+
+            # Define the device
+            dev = qml.device('default.qubit', wires=qubits)
+
+            wires = range(qubits)
+
+            @qml.qnode(dev)
+            def circuit(weights, hf_state, singles, doubles):
+                qml.templates.AllSinglesDoubles(weights, wires, hf_state, singles, doubles)
+                return qml.expval(qml.PauliZ(0))
+
+            # Evaluate the QNode for a given set of parameters
+            params = np.random.normal(0, np.pi, len(singles) + len(doubles))
+            circuit(params, hf_state, singles=singles, doubles=doubles)
+    """
+
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, weights, wires, hf_state, singles=None, doubles=None, do_queue=True):
+
+        if len(wires) < 2:
+            raise ValueError(
+                "The number of qubits (wires) can not be less than 2; got len(wires) = {}".format(
+                    len(wires)
+                )
+            )
+
+        if not singles and not doubles:
+            raise ValueError(
+                "'singles' and 'doubles' lists can not be both empty; got singles = {}, doubles = {}".format(
+                    singles, doubles
+                )
+            )
+
+        if doubles is not None:
+            for d_wires in doubles:
+                if len(d_wires) != 4:
+                    raise ValueError(
+                        "Expected entries of 'doubles' to be of size 4; got {} of length {}".format(
+                            d_wires, len(d_wires)
+                        )
+                    )
+
+        if singles is not None:
+            for s_wires in singles:
+                if len(s_wires) != 2:
+                    raise ValueError(
+                        "Expected entries of 'singles' to be of size 2; got {} of length {}".format(
+                            s_wires, len(s_wires)
+                        )
+                    )
+
+        weights_shape = qml.math.shape(weights)
+        exp_shape = self.shape(singles, doubles)
```

Comment:
You could just move the check for singles and doubles to the shape function, so that this call would fail if they are both not defined.